pkg/cimfs/cim_writer_windows.go

```diff
@@ -289,3 +345,47 @@ func GetCimUsage(ctx context.Context, cimPath string) (uint64, error) {
 	}
 	return totalUsage, nil
 }
+
+// MergeBlockCIMs creates a new merged BlockCIM from the provided source BlockCIMs.  CIM
+// at index 0 is considered to be topmost CIM and the CIM at index `length-1` is
+// considered the base CIM. (i.e file with the same path in CIM at index 0 will shadow
+// files with the same path at all other CIMs)
+// When mounting this merged CIM the source CIMs MUST be provided in the exact same order.
+func MergeBlockCIMs(mergedCIM *BlockCIM, sourceCIMs []*BlockCIM) (err error) {
+	if !IsMergedCimSupported() {
+		return fmt.Errorf("merged CIMs aren't supported on this OS version")
+	} else if len(sourceCIMs) < 2 {
+		return fmt.Errorf("need at least 2 source CIMs, got %d: %w", len(sourceCIMs), os.ErrInvalid)
+	}
+
+	var mergeFlag uint32
+	switch mergedCIM.Type {
+	case BlockCIMTypeDevice:
+		mergeFlag = CimMergeFlagBlockDevice
+	case BlockCIMTypeSingleFile:
+		mergeFlag = CimMergeFlagSingleFile
+	default:
+		return fmt.Errorf("invalid block CIM type `%d`: %w", mergedCIM.Type, os.ErrInvalid)
+	}
+
+	cim, err := CreateBlockCIM(mergedCIM.BlockPath, mergedCIM.CimName, mergedCIM.Type)
+	if err != nil {
+		return fmt.Errorf("create merged CIM: %w", err)
+	}
+	defer func() {
+		cErr := cim.Close()
+		if err == nil {
+			err = cErr
+		}
+	}()
+
+	// CimAddFsToMergedImage expects that topmost CIM is added first and the bottom
+	// most CIM is added last.
+	for _, sCIM := range sourceCIMs {
+		fullPath := filepath.Join(sCIM.BlockPath, sCIM.CimName)
+		if err := winapi.CimAddFsToMergedImage2(cim.handle, fullPath, mergeFlag); err != nil {
```

Comment:
Since we re-use `mergeFlag` here, we seem to be assuming that all CIMs will either be device or single file, but not a mix. Is this an assumption we have to make? If so, we should validate it against the input CIMs.